src/core/lib/iomgr/tcp_server_posix.cc

```diff
@@ -117,63 +130,76 @@ static grpc_error_handle CreateEventEngineListener(
             SliceBuffer* pending_data) {
           grpc_core::ApplicationCallbackExecCtx app_ctx;
           grpc_core::ExecCtx exec_ctx;
-          grpc_tcp_server_acceptor* acceptor =
-              static_cast<grpc_tcp_server_acceptor*>(
-                  gpr_malloc(sizeof(*acceptor)));
-          acceptor->from_server = s;
-          acceptor->port_index = -1;
-          acceptor->fd_index = -1;
-          if (!is_external) {
-            auto it = s->listen_fd_to_index_map.find(listener_fd);
-            if (it != s->listen_fd_to_index_map.end()) {
-              acceptor->port_index = std::get<0>(it->second);
-              acceptor->fd_index = std::get<1>(it->second);
-            }
-          } else {
-            // External connection handling.
-            grpc_resolved_address addr;
-            memset(&addr, 0, sizeof(addr));
-            addr.len = static_cast<socklen_t>(sizeof(struct sockaddr_storage));
-            // Get the fd of the socket connected to peer.
-            int fd =
-                reinterpret_cast<
-                    grpc_event_engine::experimental::PosixEndpoint*>(ep.get())
-                    ->GetWrappedFd();
-            if (getpeername(fd, reinterpret_cast<struct sockaddr*>(addr.addr),
-                            &(addr.len)) < 0) {
-              gpr_log(GPR_ERROR, "Failed getpeername: %s",
-                      grpc_core::StrError(errno).c_str());
-              close(fd);
+          grpc_pollset* read_notifier_pollset;
+          grpc_tcp_server_acceptor* acceptor;
+          void* cb_arg;
+          // Scoped for server lock, to ensure it's released before the callback
+          // is called.
+          {
+            grpc_core::MutexLockForGprMu lock(&s->mu);
```

Comment:
Does the whole block need to be under the lock or just the if(s->shutdown) statement ?

Reply:
See the lock's usage in `tcp_server_destroy` and `deactivated_all_ports`. If those can run concurrently with any of this accept logic (which includes statements like `s->listen_fd_to_index_map.find(listener_fd)`), then we have a destruction race.